### ISSUE TITLE
Adding our compiler for support in CMake

### ISSUE THREAD

**Issue by chuggafan** [LADSoft/OrangeC#603]
CMake is currently one of the more popular build systems for C++ at the moment, and one of the things we should likely do it add ourselves as a compiler version with defaults for debugging, etc. to CMake to allow for us to have a "complete" buildsystem done by them. It shouldn't be *TOO DIFFICULT*, we could even submit a PR so that we get effectively the same support OpenWatcom gets from their stuff... it's just that I have no idea how to do it...

This may or may not be actually needed to run the LLVM tests for LLVM lit....

**Comment by LADSoft** [comment]
if it turns out it is needed I can go ahead and do it; otherwise I'll push it off to the next milestone...

**Comment by LADSoft** [comment]
from the LIT documentation:


> These programs are compiled using a user specified compiler and set of flags, and then executed to capture the program output and timing information. The output of these programs is compared to a reference output to ensure that the program is being compiled correctly.

so we shouldn't need to do this immediately.    Hopefully will have the disposition of this soon.

**Comment by LADSoft** [comment]
this is next up on my list...   stay tuned...

**Comment by LADSoft** [comment]
I've submitted a patch for this...

**Comment by GitMensch** [comment]
Nice, do you mind sharing the link?

**Comment by chuggafan** [comment]
> Nice, do you mind sharing the link?

Looks like it was posted here.

https://gitlab.kitware.com/cmake/cmake/-/merge_requests/8592

**Comment by LADSoft** [comment]
it is in progress, I've completed a merge request and I think they will eventually merge what I've added to the main line.    When they do I'll post more information here, and maybe update the README.md to give preliminary info for those who are adventurous enough to download from the repository.

**Comment by LADSoft** [comment]
the current status is that I worked with the CMAKE team on some issues involving how the compiler is presented...   we eventually decided to call it OrangeC (as opposed to OCC) for purposes of the cmake project.

After that they came back and said ok lets run the integration tests that clang uses to test its toolchain files and see what happens.

So I did that and out of 491 tests, 68 failed.   There were three major classes of failures:

1) incompatibilities between the way orange c and cmake handle libraries
     the file extension differs, and the fact the win32 libraries are all in one import library in orange C. throws a monkey wrench in
2) incompatibilities in the toolchain files
    for example cmake tests should build a win32 gui file properly, but that doesn't work for orange C due I assume to missing CMAKE rules...   I don't know the full extent of the problem though...
3) compiler crashes
    this was rare but happened once or twice

I'm going to be looking at #2 and #3 as time permits...   #1 will likely not work without a rewrite of the tests in question...

**Comment by LADSoft** [comment]
Compilers known by cmake get nightly testing with their test suite; I've been working with that and found a few bugs in Orange C.

So I had another look and found we had fixed all the bugs that that the tests they wanted me to run made obvious in Orange C.    There were still many tests failing apparently due to mismatches between the compiler and the tests, and so I sent a message on to the CMAKE team.    I have yet to confirm but it looks like there is going to be some combination of excluding tests that are no longer relevant, and working on updating the tests to be inclusive for Orange C.   I'm not sure yet who they envision doing those updates although I'm suspecting they are going to want me to.    

This was going to be a long-cycle project anyway as they are in the middle of a release and I don't know what their schedule for the next one is...   

stay tuned for more info.

**Comment by LADSoft** [comment]
I finally got enough ducks in a row that we could have a release...   about two days ago though the cmake team got tired of waiting on me and closed the pull request.   I think they can open it again though....   anyway after the release made it to this board I went and left a note in the issue (it is 25032 in their system) that I'd had a release and we can progress again...

**Comment by LADSoft** [comment]
they ran an initial set of tests today and came back with a bug:   orange c won't let you do exports from an EXE file.    That is halting a lot of the tests they do...   I'm going to fix that over the weekend.

I'm also going to be changing the versioning to use major, minor, build like everyone else does in the compiler world.....

**Comment by LADSoft** [comment]
they finished the first pass of the tests last wednesday; there were a total of six or eight minor bugs found in OCC.   I've got them worked out and will submit this week...

**Comment by GitMensch** [comment]
So there's a release this week?

**Comment by LADSoft** [comment]
no im going to submit the next round of changes and see where we are now :).    There may be a little bit of a delay before they get back to me though...

**Comment by LADSoft** [comment]
the current status is that I submitted a new version of the compiler; there is a small amount of rework to be done but nothing too difficult.   We are getting close to the end on this...

**Comment by LADSoft** [comment]
so curent status is orangec passed all their tests.    I'm evaluating whether we can have another release at this point....   main problem with a release being I'd been having ongoing problems with my alternate branches just randomly crashing during builds.    But I may have fixed that now...   have put everything on master and assuming the current test succeeds I will push it to see what appveyor does with it.

**Comment by LADSoft** [comment]
so current status on this is when we tried to rerun the tests with an occ built with itself, one of the tests failed.   I spent this week discovering that sqlite3 has a bug that sometimes affected generation of debug files, so I updated to the latest and the problem went away.    Getting ready to do a new build...

**Comment by LADSoft** [comment]
OCC Passed their tests now.   From what I understand after the current release, they will start running orange c version 6.73.1 as part of their nightly regression testing then it should be part of the next release after the current one.

**Comment by LADSoft** [comment]
So I finally got around to catching up and verifying the cmake holds the orange c compiler configuration files.    I downloaded CMAKE 3.29.3 and can verify they are in that version; they may have been in 3.28 as well, I am unsure on that.
